File: app/src/main/java/com/yeraygarcia/recipes/database/remote/Webservice.java
package com.yeraygarcia.recipes.database.remote;

import android.arch.lifecycle.LiveData;

import com.yeraygarcia.recipes.database.entity.Recipe;
import com.yeraygarcia.recipes.database.entity.RecipeIngredient;
import com.yeraygarcia.recipes.database.entity.ShoppingListItem;
import com.yeraygarcia.recipes.database.entity.custom.All;

import java.util.List;

import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.Headers;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Query;

public interface Webservice {

    @POST("tokensignin")
    @Headers("No-Authentication: true")
    Call<ResourceData<User>> postToken(@Query("id_token") String token);

    @GET("all")
    LiveData<ApiResponse<ResourceData<All>>> getAll();

    @GET("recipes")
    LiveData<ApiResponse<ResourceData<List<Recipe>>>> getRecipes();

    @POST("shopping-list-items")
    LiveData<ApiResponse<ResourceData<ShoppingListItem>>> postShoppingListItem(@Body ShoppingListItem item);

    @PUT("recipe-ingredients")
    Call<ResourceData<RecipeIngredient>> updateRecipeIngredient(@Body RecipeIngredient recipeIngredient);
}

File: app/src/main/java/com/yeraygarcia/recipes/database/remote/Webservice.kt
package com.yeraygarcia.recipes.database.remote

import android.arch.lifecycle.LiveData
import com.yeraygarcia.recipes.database.entity.Ingredient
import com.yeraygarcia.recipes.database.entity.Recipe
import com.yeraygarcia.recipes.database.entity.RecipeIngredient
import com.yeraygarcia.recipes.database.entity.RecipeStep
import com.yeraygarcia.recipes.database.entity.custom.All
import com.yeraygarcia.recipes.database.entity.custom.SyncDto
import retrofit2.Call
import retrofit2.http.*

interface Webservice {

    @POST("tokensignin")
    @Headers("No-Authentication: true")
    fun postToken(@Query("id_token") token: String): Call<ResourceData<User>>

    @GET("all")
    fun getAll(): LiveData<ApiResponse<ResourceData<All>>>

    @PUT("recipes")
    fun updateRecipe(@Body recipe: Recipe): Call<ResourceData<Recipe>>

    @PUT("recipe-ingredients")
    fun updateRecipeIngredient(@Body recipeIngredient: RecipeIngredient): Call<ResourceData<RecipeIngredient>>

    @PUT("recipe-steps")
    fun updateRecipeStep(@Body recipeStep: RecipeStep): Call<ResourceData<RecipeStep>>

    @POST("ingredients")
    fun insertIngredient(@Body ingredient: Ingredient): Call<ResourceData<Ingredient>>

    @POST("recipe-ingredients")
    fun insertRecipeIngredient(recipeIngredient: RecipeIngredient): Call<ResourceData<RecipeIngredient>>

    @POST("sync")
    fun sync(@Body dataToSync: SyncDto): Call<ResourceData<SyncDto>>
}
